core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java

```diff
@@ -138,16 +140,18 @@ public String toString() {
   private void updateGlobalWatermarkIfNeeded(Context context) {
     long now = System.currentTimeMillis();
     if (now - globalWatermarkUpdatePeriodMs > lastGlobalWatermarkUpdate) {
-      tracker.update(processName, context.getMinWatermark());
+      CompletableFuture<Void> update = tracker.update(processName, context.getMinWatermark());
+      if (lastGlobalWatermarkUpdate == Long.MIN_VALUE) {
+        // when there was no update yet, wait for the update to happen
+        ExceptionUtils.ignoringInterrupted(update::get);
+      }
       lastGlobalWatermarkUpdate = now;
-      log.debug("Updated watermark of {} to {}", this, globalWatermark);
     }
-    globalWatermark = tracker.getGlobalWatermark(processName, context.getMinWatermark());
   }
 
   @VisibleForTesting
   void forceUpdateGlobalWatermark() {
-    lastGlobalWatermarkUpdate = 0;
+    lastGlobalWatermarkUpdate = Long.MIN_VALUE;
```

Comment:
This seems like a workaround for not using `Clock` instead of `System.currentTimeMillis()` ...

Reply:
I would not use `Clock`. We could create a `TimeProvider` in core, but that seems like overkill for what we need here.